File: app/src/main/java/es/etologic/mahjongscoring2/app/base/BaseViewModel.java
package es.etologic.mahjongscoring2.app.base;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import es.etologic.mahjongscoring2.app.model.ShowState;
import io.reactivex.disposables.CompositeDisposable;

public class BaseViewModel extends ViewModel {

    protected CompositeDisposable disposables = new CompositeDisposable();
    protected MutableLiveData<ShowState> progressState = new MutableLiveData<>();
    protected MutableLiveData<String> snackbarMessage = new MutableLiveData<>();
    protected MutableLiveData<Throwable> error = new MutableLiveData<>();

    public LiveData<ShowState> getLocalProgressState() { return progressState; }
    public LiveData<String> getSnackbarMessage() { return snackbarMessage; }
    public LiveData<Throwable> getError() { return error; }

    @Override protected void onCleared() {
        if(!disposables.isDisposed()) {
            disposables.dispose();
        }
        super.onCleared();
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/base/BaseViewModel.kt
package es.etologic.mahjongscoring2.app.base

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import es.etologic.mahjongscoring2.app.model.ShowState
import es.etologic.mahjongscoring2.domain.model.Round
import io.reactivex.disposables.CompositeDisposable

open class BaseViewModel : ViewModel() {
    
    protected var disposables = CompositeDisposable()
    protected var progressState = MutableLiveData<ShowState>()
    protected var snackbarMessage = MutableLiveData<String>()
    protected var error = MutableLiveData<Throwable>()
    
    internal fun getProgressState(): LiveData<ShowState> = progressState
    internal fun getSnackbarMessage(): LiveData<String> = snackbarMessage
    internal fun getError(): LiveData<Throwable> = error
    
    override fun onCleared() {
        if (!disposables.isDisposed)
            disposables.dispose()
        super.onCleared()
    }
}
